tls/resource_cert_revocation_list.go

```diff
@@ -0,0 +1,193 @@
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceCertRevocationList() *schema.Resource {
+	s := map[string]*schema.Schema{
+		"certs_to_revoke": &schema.Schema{
+			Type: schema.TypeList,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+			Required:    true,
+			Description: "PEM-encoded certificates to be revoked",
+			ForceNew:    true,
+			StateFunc: func(v interface{}) string {
+				return hashForState(strings.Join(v.([]string), ","))
+			},
+		},
+		"early_renewal_hours": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Default:     0,
+			Description: "Number of hours before the certificates expiry when a new certificate will be generated",
+		},
+		"validity_period_hours": {
+			Type:        schema.TypeInt,
+			Required:    true,
+			Description: "Number of hours that the CRL will remain valid for",
+			ForceNew:    true,
+		},
+		"validity_start_time": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+
+		"validity_end_time": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"ready_for_renewal": {
+			Type:     schema.TypeBool,
+			Computed: true,
+		},
+		"crl_pem": {
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"ca_cert_pem": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "PEM-encoded CA certificate",
+			ForceNew:    true,
+			StateFunc: func(v interface{}) string {
+				return hashForState(v.(string))
+			},
+		},
+		"ca_private_key_pem": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "PEM-encoded CA private key used to sign the CRL",
+			ForceNew:    true,
+			Sensitive:   true,
+			StateFunc: func(v interface{}) string {
+				return hashForState(v.(string))
+			},
+		},
+		"ca_key_algorithm": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Name of the algorithm used to generate the certificate's private key",
+			ForceNew:    true,
+		},
+	}
+
+	return &schema.Resource{
+		Create:        CreateCRL,
+		Delete:        DeleteCRL,
+		Read:          ReadCRL,
+		Update:        UpdateCRL,
+		CustomizeDiff: CustomizeCertificateDiff,
+		Schema:        s,
+	}
+}
+
+func CreateCRL(d *schema.ResourceData, meta interface{}) error {
+	certsToRevoke := make([]pkix.RevokedCertificate, 0)
+	if value := d.Get("certs_to_revoke").([]interface{}); len(value) > 0 {
+		for _, vi := range value {
+			block, err := decodePEMFromBytes([]byte(vi.(string)), "")
+			if err != nil {
+				return err
+			}
+			certificate, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return err
+			}
+			certsToRevoke = append(certsToRevoke, pkix.RevokedCertificate{
+				SerialNumber: certificate.SerialNumber,
+			})
+		}
+	}
+	caKey, err := parsePrivateKey(d, "ca_private_key_pem", "ca_key_algorithm")
+	if err != nil {
+		return err
```

Comment:
I think it would be useful to annotate those errors using `fmt.Errorf`, so the error message is more clear for the user:
```suggestion
		return fmt.Errorf("failed to parse %q field: %w", err)
```

Also for other `return err`, which faces user.

Reply:
I can do that very easily, I just followed the same error handling in other parts of the code that also do this private key parsing, for consistency with the rest of the code. Should I change it here anyway? I can also change it in the other parts a key or certificate is parsed.

Reply:
In my opinion, ideally all the errors should be wrapped, as far as I know it's considered a good practice (https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-
gracefully) as it allows to quickly identify what exactly broke and provide better context for the user.

The comment I left was just an example on one line (pointing out every `return err` would be a lot of work :smile:), so for any new code, I'd try to follow this. I just wouldn't change the existing codebase, as this should be done in separate context (PR/issue). I hope that make sense.

Reply:
Agree, done! wrapped the errors of the new code in this PR